internal/services/v1/bulkcheck.go

```diff
@@ -0,0 +1,249 @@
+			return err
+		}
+
+		bulkResponseMutex.Lock()
+		defer bulkResponseMutex.Unlock()
+
+		for _, resourceID := range resourceIDs {
+			reqItem, err := requestItemFromResourceAndParameters(params, resourceID)
+			if err != nil {
+				return bc.rewriteError(ctx, err)
+			}
+
+			if err := addPair(&v1.CheckBulkPermissionsPair{
+				Request: reqItem,
+				Response: &v1.CheckBulkPermissionsPair_Error{
+					Error: statusResp.Proto(),
+				},
+			}); err != nil {
+				return bc.rewriteError(ctx, err)
+			}
+		}
+
+		return nil
+	}
+
+	appendResultsForCheck := func(params *computed.CheckParameters, resourceIDs []string, metadata *dispatchv1.ResponseMeta, results map[string]*dispatchv1.ResourceCheckResult) error {
+		bulkResponseMutex.Lock()
+		defer bulkResponseMutex.Unlock()
+
+		for _, resourceID := range resourceIDs {
+			reqItem, err := requestItemFromResourceAndParameters(params, resourceID)
+			if err != nil {
+				return bc.rewriteError(ctx, err)
+			}
+
+			if err := addPair(&v1.CheckBulkPermissionsPair{
+				Request:  reqItem,
+				Response: pairItemFromCheckResult(results[resourceID]),
+			}); err != nil {
+				return bc.rewriteError(ctx, err)
+			}
+		}
+
+		respMetadata.DispatchCount += metadata.DispatchCount
+		respMetadata.CachedDispatchCount += metadata.CachedDispatchCount
+		return nil
+	}
+
+	for _, group := range groupedItems {
+		group := group
+
+		slicez.ForEachChunk(group.resourceIDs, MaxBulkCheckDispatchChunkSize, func(resourceIDs []string) {
+			tr.Add(func(ctx context.Context) error {
+				ds := datastoremw.MustFromContext(ctx).SnapshotReader(atRevision)
+
+				// Ensure the check namespaces and relations are valid.
+				err := namespace.CheckNamespaceAndRelations(ctx,
+					[]namespace.TypeAndRelationToCheck{
+						{
+							NamespaceName: group.params.ResourceType.Namespace,
+							RelationName:  group.params.ResourceType.Relation,
+							AllowEllipsis: false,
+						},
+						{
+							NamespaceName: group.params.Subject.Namespace,
+							RelationName:  stringz.DefaultEmpty(group.params.Subject.Relation, graph.Ellipsis),
+							AllowEllipsis: true,
+						},
+					}, ds)
+				if err != nil {
+					return appendResultsForError(group.params, resourceIDs, err)
+				}
+
+				// Call bulk check to compute the check result(s) for the resource ID(s).
+				rcr, metadata, err := computed.ComputeBulkCheck(ctx, bc.dispatch, *group.params, resourceIDs)
+				if err != nil {
+					return appendResultsForError(group.params, resourceIDs, err)
+				}
+
+				return appendResultsForCheck(group.params, resourceIDs, metadata, rcr)
+			})
+		})
+	}
+
+	// Run the checks in parallel.
+	if err := tr.StartAndWait(); err != nil {
+		return nil, bc.rewriteError(ctx, err)
+	}
+
+	return &v1.CheckBulkPermissionsResponse{CheckedAt: checkedAt, Pairs: orderedPairs}, nil
+}
+
+func toCheckBulkPermissionsRequest(req *v1.BulkCheckPermissionRequest) *v1.CheckBulkPermissionsRequest {
+	items := make([]*v1.CheckBulkPermissionsRequestItem, len(req.Items))
```

Comment:
you could also do `, 0, len(req.Items)` and `append`, but this works

Reply:
This was a direct copy-paste from the previous implementation